src/tpm2_pytss/TCTISPIHelper.py

```diff
@@ -0,0 +1,261 @@
+# SPDX-License-Identifier: BSD-2
+
+from .internal.utils import _chkrc, _lib_version_atleast
+from ._libtpm2_pytss import ffi, lib
+from .constants import TSS2_RC, TPM2_RC
+from .TSS2_Exception import TSS2_Exception
+from .TCTI import TCTI
+
+if not _lib_version_atleast("tss2-tcti-spi-helper", "0.0.0"):
+    raise NotImplementedError("Package tss2-tcti-spi-helper not present")
+
+
+@ffi.def_extern()
+def _tcti_spi_helper_sleep_ms(userdata, milliseconds):
+    thiz = TCTISPIHelper._cffi_cast(userdata)
+    if not hasattr(thiz, "on_sleep_ms"):
+        thiz.on_sleep_ms(milliseconds)
+
```

Comment:
Might be good to catch any exception, otherwise cffi might just dump the traceback to stderr

Reply:
> Might be good to catch any exception, otherwise cffi might just dump the traceback to stderr

That's actually what I wanted. Theirs no mechanism to return errors through the C API, which makes me want to change that before I cut 4.0 of tpm2-tss. The SPI TCTI Helper just retries 100 times and fails with IO error. It would be better to have the ability to error in all the spi callbacks and then we can set last exception and re-throw it in the Python code.

Reply:
One option is to simply catch and pass, no feedback in case of errors, but no random traceback dumps either

Reply:
@whooo 
> One option is to simply catch and pass, no feedback in case of errors, but no random traceback dumps either

I don't get the benefit of that, could you explain? It's moot now, as I updated the API to use return codes everywhere but finalize, however I am curious.

Now the exception framework of the base TCTI class can be used through the set_last_exception and friends methods and then we can also check it in the constructor for the spi helper TCTI and catch errors as well. This allows for exceptions to be handled in the constructor as well as in the TCTI interface calls, like transmit.